src/middleware.js

```diff
@@ -15,6 +14,160 @@ import { handleCarBlock } from './handlers/car-block.js'
+  if (!rateLimiter) {
+    console.warn('no rate limiter found')
+    return RATE_LIMIT_EXCEEDED.NO
+  }
+  const rateLimitResponse = await rateLimiter.limit({ key: cid.toString() })
+  if (rateLimitResponse.success) {
+    return RATE_LIMIT_EXCEEDED.NO
+  } else {
+    console.log(`limiting CID ${cid}`)
+    return RATE_LIMIT_EXCEEDED.YES
+  }
+}
+
+/**
+ *
+ * @param {Environment} env
+ * @param {string} authToken
+ * @returns TokenMetadata
+ */
+async function getTokenMetadata (env, authToken) {
+  const cachedValue = await env.AUTH_TOKEN_METADATA.get(authToken)
+  // TODO: we should implement an SWR pattern here - record an expiry in the metadata and if the expiry has passed, re-validate the cache after
+  // returning the value
+  if (cachedValue) {
+    return deserializeTokenMetadata(cachedValue)
+  } else {
+    const accounting = Accounting.create({ serviceURL: env.ACCOUNTING_SERVICE_URL })
+    const tokenMetadata = await accounting.getTokenMetadata(authToken)
+    if (tokenMetadata) {
+      await env.AUTH_TOKEN_METADATA.put(authToken, serializeTokenMetadata(tokenMetadata))
+      return tokenMetadata
+    } else {
+      return null
+    }
+  }
+}
+
+/**
+ * @type {import('./bindings.js').RateLimits}
+ */
+const RateLimits = {
+  create: ({ env }) => ({
+    check: async (cid, options) => {
+      const authToken = await getAuthorizationTokenFromRequest(options)
+      if (authToken) {
+        console.log(`found token ${authToken}, looking for content commitment`)
+        const tokenMetadata = await getTokenMetadata(env, authToken)
+
+        if (tokenMetadata) {
+          if (tokenMetadata.invalid) {
+            // this means we know about the token and we know it's invalid, so we should just use the CID rate limit
+            return checkRateLimitForCID(env, cid)
+          } else {
+            // TODO at some point we should enforce user configurable rate limits and origin matching
+            // but for now we just serve all valid token requests
+            return RATE_LIMIT_EXCEEDED.NO
+          }
+        } else {
+          // we didn't get any metadata - for now just use the top level rate limit
+          // this means token based requests will be subject to normal rate limits until the data propagates
+          return checkRateLimitForCID(env, cid)
+        }
+      } else {
+        // no token, use normal rate limit
+        return checkRateLimitForCID(env, cid)
+      }
+    }
+  })
+}
+
+/**
+ * @type {import('./bindings.js').Accounting}
+ */
+const Accounting = {
+  create: ({ serviceURL }) => ({
+    record: async (cid, options) => {
+      console.log(`using ${serviceURL} to record a GET for ${cid} with options`, options)
+    },
+
+    getTokenMetadata: async () => {
+      // TODO I think this needs to check the content claims service (?) for any claims relevant to this token
+      // TODO do we have a plan for this? need to ask Hannah if the indexing service covers this?
```

Comment:
Yes eventually this will be indexing service, but you may have to use content claims temporarily. Either way I believe the current idea is for them to have the same API.